### PR TITLE
Exit only when report file stream ended to avoid partial file

### PR DESCRIPTION
**Pull request by lowik**
The process was closed before the file report stream and so the report file was not completed.
The result with Jenkins TAP plugin is:
```
org.tap4j.parser.ParserException: Error parsing TAP Stream: Missing TAP Plan.
```

### CODE REVIEW

File: lib/app.js [hunk 1]
```diff
@@ -45,6 +38,25 @@ function App(config, finalizer) {
...
+        exitCode = 1;
+      }
+      (finalizer || cleanExit)(exitCode, err);
+    };
+
+    if (this.reportFile) {
+      // Exit only when report file stream ended
+      this.reportFile.fileStream.on('finish', exit);
+      this.reportFile.fileStream.on('close', exit);
+      this.reportFile.fileStream.on('error', exit);
```
Comment (johanneswuerbach) [inline lib/app.js]: This requires a guard to not call `exit` multiple times.

File: lib/report_file.js [hunk 2]
```diff
@@ -17,10 +17,13 @@ function ReportFile(reportFile, out) {
...
-  this.outputStream.on('end', function(data) {
-    out.end(data);
+  function finish(data) {
     fileStream.end(data);
-  });
+  }
+
+  this.outputStream.on('finish', finish);
+  this.outputStream.on('close', finish);
+  this.outputStream.on('error', finish);
```
Comment (johanneswuerbach) [inline lib/report_file.js]: Same ^^^

**Comment by johanneswuerbach** [comment]
Thanks for doing that, that was one reason for flaky tests I couldn't solve yet :+1: 

Two nits, otherwise LGTM.

**Comment by johanneswuerbach** [comment]
Sorry, this needs another rebase :-)

**Comment by lowik** [comment]
Rebase done

**Comment by johanneswuerbach** [comment]
Looks like the test failure is related, mind taking a look?

**Comment by lowik** [comment]
Strange, the tests works fine at home and for one on appveyor.

I had this error before fixing tests/ci/report_file_tests.js.
Seems like the fix is not taken into account.

I add log checking that

**Comment by lowik** [comment]
No log "entering app finalizer callback" on travis build

**Comment by johanneswuerbach** [comment]
Hey, thanks for the PR. I fixed the broken test and rebased your commits here https://github.com/testem/testem/issues/770 :-)

**Pull request closed without merging**